ksql-engine/src/main/java/io/confluent/ksql/engine/EngineContext.java

```diff
@@ -55,42 +59,49 @@
   private final DdlCommandExec ddlCommandExec;
   private final QueryIdGenerator queryIdGenerator;
   private final ProcessingLogContext processingLogContext;
-  private final KsqlParser parser = new DefaultKsqlParser();
+  private final KsqlParser parser;
   private final BiConsumer<ServiceContext, QueryMetadata> outerOnQueryCloseCallback;
   private final Map<QueryId, PersistentQueryMetadata> persistentQueries;
+  private final StatementRewriteForStruct rewriter;
 
-  private EngineContext(
+  static EngineContext create(
       final ServiceContext serviceContext,
       final ProcessingLogContext processingLogContext,
       final MutableMetaStore metaStore,
       final QueryIdGenerator queryIdGenerator,
       final BiConsumer<ServiceContext, QueryMetadata> onQueryCloseCallback
   ) {
-    this.serviceContext = Objects.requireNonNull(serviceContext, "serviceContext");
-    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
-    this.queryIdGenerator = Objects.requireNonNull(queryIdGenerator, "queryIdGenerator");
-    this.ddlCommandFactory = new CommandFactories(serviceContext, metaStore);
-    this.outerOnQueryCloseCallback = Objects
-        .requireNonNull(onQueryCloseCallback, "onQueryCloseCallback");
-    this.ddlCommandExec = new DdlCommandExec(metaStore);
-    this.persistentQueries = new ConcurrentHashMap<>();
-    this.processingLogContext = Objects
-        .requireNonNull(processingLogContext, "processingLogContext");
+    return new EngineContext(
+        serviceContext,
+        processingLogContext,
+        metaStore,
+        queryIdGenerator,
+        onQueryCloseCallback,
+        new DefaultKsqlParser(),
+        new StatementRewriteForStruct()
+    );
   }
 
-  static EngineContext create(
+  @VisibleForTesting
+  EngineContext(
       final ServiceContext serviceContext,
       final ProcessingLogContext processingLogContext,
       final MutableMetaStore metaStore,
       final QueryIdGenerator queryIdGenerator,
-      final BiConsumer<ServiceContext, QueryMetadata> onQueryCloseCallback
+      final BiConsumer<ServiceContext, QueryMetadata> onQueryCloseCallback,
+      final KsqlParser parser,
+      final StatementRewriteForStruct rewriter
```

Comment:
Should we make this a list of rewriters? We could also have that include the `ROWTIME` rewrite (though this would require adding support for rewriting the whole statement to `StatementRewriteForRowtime`). However, even if we don't do that I think a collection of rewriters is what we want here.

Reply:
I did think about adding a `StatementRewriter` interface. We can then add a composite implementation that is itself a list of rewriters, i.e. `EngineContext` doesn't need to worry about the list of rewriters, just a single one.

However, I didn't make the change as its not required yet.  Let's worry about it once we have more than one re-writer.

As you say, we could change `StatementRewriteForRowtime ` to always be called, but I saw that as out of scope for this change.